proxy/proxy/client.go

```diff
@@ -323,3 +320,23 @@ func NewConnSrc(instance string, l net.Listener) <-chan Conn {
 	}()
 	return ch
 }
+
+// Shutdown waits up to a given amount of time for all active connections to
+// close. No new connections can be established after calling this method.
+// Returns an error if there are still active connections after waiting for the
+// whole length of the timeout.
+func (c *Client) Shutdown(termTimeout time.Duration) error {
+	// Don't allow new connections while terminating.
+	atomic.StoreUint64(&c.MaxConnections, 0)
```

Comment:
Wait but... `MaxConnections == 0` implies there should be no limit on the number of connections, so this doesn't actually stop new connections. We can't use -1 because the type is Uint :( Maybe we should create a private sentinel value of the max uint64 value and use that to indicate "block new connections"; I doubt someone is going to actually set the max number of connections to a value that high so I don't think it'll cause any real problems. (famous last words...)